File: bindings/typescript/src/assemblers/text.ts
import { FragmentId, TextPrimitiveFragment } from '@vizstack/schema';
import { FragmentAssembler } from '../fragment-assembler';

type VariantType = TextPrimitiveFragment['contents']['variant'];
type EmphasisType = TextPrimitiveFragment['contents']['emphasis'];

class TextPrimitiveFragmentAssembler extends FragmentAssembler {
    private _text: string;
    private _variant?: VariantType;
    private _emphasis?: EmphasisType;

    constructor(
        text: string,
        variant?: VariantType,
        emphasis?: EmphasisType,
    ) {
        super();
        this._text = text;
        this._variant = variant;
        this._emphasis = emphasis;
    }

    public assemble(getId: (obj: any, name: string) => FragmentId): [TextPrimitiveFragment, any[]] {
        return [
            {
                type: 'TextPrimitive',
                contents: {
                    text: this._text,
                    variant: this._variant,
                    emphasis: this._emphasis,
                },
                meta: this._meta,
            },
            [],
        ];
    }
}

export function Text(
    text: string,
    variant?: VariantType,
    emphasis?: EmphasisType,
) {
    return new TextPrimitiveFragmentAssembler(text, variant, emphasis);
}

export interface Text extends ReturnType<typeof Text> {}

File: bindings/python/vizstack/assemblers/text.py
from vizstack.fragment_assembler import FragmentAssembler
from typing import Optional, Tuple, Dict, List, Any
from typing_extensions import Literal
from vizstack.schema import JsonType, View, Fragment

Variant = Literal['caption', 'body',  'subheading', 'heading', None]
Emphasis = Literal['normal', 'less', 'more', None]


class Text(FragmentAssembler):
    """
    A View which renders a contiguous block of text.
    """

    def __init__(self,
                 text: str,
                 variant: Variant = None,
                 emphasis: Emphasis = None,) -> None:
        """
        Args:
            text: Text which should be rendered.
            variant: Semantic role of text.
            emphasis: Relative important of text.
        """
        super(Text, self).__init__()
        self._text: str = text
        self._variant: Variant = variant
        self._emphasis: Emphasis = emphasis

    def assemble(self, get_id) -> Tuple[Fragment, List[Any]]:
        return {
            'type': 'TextPrimitive',
            'contents': {
                'text': self._text,
                'variant': self._variant,
                'emphasis': self._emphasis,
            },
            'meta': self._meta,
        }, []
